Add item requests, microservice,  tests

KanID92 opened the pull request.

KanID92 committed:
- исправить поля Item на объектную модель;
- добавить обработку запросов на вещи.

KanID92 committed:
- рефакторинг. Разбить на 2 приложения: ShareItServer, ShareItGateway

KanID92 committed:
- добавить тесты;

KanID92 committed:
- откорректировать по ошибкам checkstyle;

KanID92 committed:
- откорректировать по ошибкам checkstyle;

kesch9 commented on gateway/pom.xml:
```diff
@@ -0,0 +1,70 @@
+
+        <dependency>
+            <groupId>org.hibernate.validator</groupId>
+            <artifactId>hibernate-validator</artifactId>
```
hibernate-validator

Действительно ли требуется эта зависимость?)
Выше уже присутствует spring-boot-starter-validation, которая включает в себя все необходимое)

kesch9 commented on gateway/pom.xml:
```diff
@@ -0,0 +1,70 @@
+
+        <dependency>
+            <groupId>org.springframework.boot</groupId>
+            <artifactId>spring-boot-starter-test</artifactId>
```
Тестов нет, поэтому эта зависимость излишняя

kesch9 commented on gateway/src/main/java/ru/practicum/shareit/booking/BookingController.java:
```diff
@@ -0,0 +1,80 @@
+	@GetMapping("/owner")
+	public ResponseEntity<Object> getCurrentBookingsByOwnerId(
+			@RequestHeader(USER_ID_REQUEST_HEADER) long ownerUserId,
+			@RequestParam(name = "status", required = false) String state) {
```
Обязательно
Нужно провалидировать state в gateway, он должен соответствовать enum. А в сервере сразу принимать SearchState, вместо строки. Это замечание и для других enum, которые используются в запросах.

kesch9 replied:
Подправлено не верно, на сервере принимаем enum, в gateway строку и добавляем default значение. Там же в gateway нужно провалидировать строку. У вас все наоборот сделано, в gateway принимается enum, а на сервере строка плюс валидация

kesch9 commented on gateway/src/main/java/ru/practicum/shareit/handler/ErrResponse.java:
```diff
@@ -0,0 +1,14 @@
+    String error;
+    String description;
+
+    public ErrResponse(String error, String description) {
```
Можно создать конструктор через ломбок @AllArgsConstructor

kesch9 commented on gateway/src/main/java/ru/practicum/shareit/item/dto/ItemUpdateDto.java:
```diff
@@ -10,8 +9,8 @@
 
-    @NotNull(message = "Updated Item must have id")
-    long id;
+//    @NotNull(message = "Updated Item must have id")
```
Закоментированный код надо удалять на ревью

kesch9 commented on gateway/src/main/java/ru/practicum/shareit/request/ItemRequestController.java:
```diff
@@ -0,0 +1,76 @@
+    public ResponseEntity<Object> getAllRequests(
+            @RequestHeader(name = USER_ID_REQUEST_HEADER) long userId,
+            @RequestParam int from,
+            @RequestParam Integer size) {
```
Обязательно
from и size надо также провалидировать
@PositiveOrZero @RequestParam(name = "from", defaultValue = "0") Integer from,
			@Positive @RequestParam(name = "size", defaultValue = "10") Integer size

kesch9 commented on server/src/main/java/ru/practicum/shareit/request/service/ItemRequestServiceImpl.java:
```diff
@@ -0,0 +1,94 @@
+    public List<ItemRequestOutputDto> getAllRequests(long userId, int from, int size) {
+        userRepository.findById(userId)
+                .orElseThrow(() -> new NotFoundException("User with id " + userId + " not found"));
+        return itemRequestRepository.findItemRequestWithFromAndSize(from, size)
```
Обязательно
Нужно использовать стандартную пагинацию спринга, вместо написание запроса
https://www.baeldung.com/spring-data-jpa-pagination-sorting

kesch9 reviewed (commented):
Нужно немного подправить работу

KanID92 committed:
fix:
- убрать закомментированный код;
- заменить String в аргументах контроллеров на Enum;
- ErrResponse. Заменить создание конструктора аннотацией Lombok;
- ItemRequestController. Добавить валидацию в методах для аргументов from и size;
- Модуль gateway. pom.xml. Убрать лишние зависимости hibernate-validator, spring-boot-starter-test.

KanID92 committed:
fix:
- откорректировать ошибку Enum SearchState.

KanID92 committed:
fix:
- откорректировать ошибку Enum SearchState.

KanID92 committed:
fix:
- откорректированы тесты в соответствии с пагинацией.

kesch9 reviewed (approved):
Approve
